### DISCUSSION TITLE
Kraken KeyError on pair

### DISCUSSION THREAD

**Discussion by melewitz** [bmoscon/cryptofeed#396] [Q&A] [answered]
I'm new to cryptofeed and really love the framework.  I hope I'm just doing something wrong here...

I see repeated exceptions when running the simple snippet (shown below) that was basically copied from the demo_book_delta example.  I suspect this occurs when an L2 delta update arrives before a full book has been received.  If so, shouldn't the library protect against that?  Should I wait and add the L2 delta subscription after I've seen a book for the pair?

Exception:
```
2021-01-15 16:06:23,097 : ERROR : KRAKEN-a61789: encountered an exception, reconnecting
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/cryptofeed/feedhandler.py", line 230, in _connect
    await self._handler(connection, handler)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/cryptofeed/feedhandler.py", line 259, in _handler
    await handler(message, connection, self.last_msg[connection.uuid])
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/cryptofeed/exchange/kraken.py", line 162, in message_handler
    await self._book(msg, pair, timestamp)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/cryptofeed/exchange/kraken.py", line 135, in _book
    self.l2_book[pair][side][price] = size
KeyError: 'MANA-BTC'
```

Code:
```
class OrderBook(object):
    def __init__(self, name):
        self.L2 = {
            L2_BOOK: BookCallback(self.handle_book),
            BOOK_DELTA: BookUpdateCallback(self.handle_l2_delta)
        }
    
    <snip empty methods for handle_book and handle_ls_delta that just return>

def main():
    pairs = [ <snip> ]
    f = FeedHandler()
    f.add_feed(Kraken(max_depth=50, book_interval=10, pairs=pairs, channels=[L2_BOOK], callbacks=OrderBook("Kraken").L2))
    f.run()
```

**Comment by bmoscon** [comment]
The example you are using does not work the way you think it works. None of them will work with multiple pairs. Do not use that as an example to base normal book building off of - its sole purpose is to verify that deltas in general work for an exchange.

**Reply by melewitz** [reply]
Thanks for the quick response.  Is there an example of a good way to monitor multiple pairs, or documentation to get me thinking about it correctly?

**Comment by bmoscon** [comment]
the code in examples\demo.py configures multiple pairs on several exchanges for book updates.

**Comment by bmoscon** [comment, marked as answer]
to follow back up on this - changes were made to kraken to handle situations similar to these when subscribing to large quantities of symbols
